### PR TITLE
params: manage access to ssl private keys

### PR DESCRIPTION
**Pull request by josephholsten**

### CODE REVIEW

**Comment by mmoll** [comment]
@josephholsten Puppet does manage (some) things on its own, that why this module generates e.g. for me in puppet.conf:
```ini
[main]
    hostprivkey = $privatekeydir/$certname.pem { mode = 640 }
    privatekeydir = $ssldir/private_keys { group = service }
```
I didn't dig further, but you'd need to tackle your needs on that level.

File: spec/classes/puppet_server_config_spec.rb [hunk 1]
```diff
@@ -187,6 +187,18 @@
         end
       end
 
+      describe 'with server_private_keys_group' do
+        let :pre_condition do
+          "class {'puppet': server => true; server_private_keys_group => 'heyting'}"
```
Comment (mmoll) [inline spec/classes/puppet_server_config_spec.rb]: comma instead of semicolon

**Comment by josephholsten** [comment]
well I'm very confused as to why the test is getting nil instead of the default `$::puppet::group`

File: manifests/params.pp [hunk 2]
```diff
@@ -266,6 +266,10 @@
   $server_environments_owner   = $user
   $server_environments_group   = $root_group
   $server_environments_mode    = '0755'
+  # Owner of the private_keys dir: for cases where external services need read
+  # access to it.
+  $server_private_keys_owner   = $user,
+  $server_private_keys_group   = $group,
```
Comment (stbenjam) [inline manifests/params.pp]: No need for commas in params.pp

Comment (josephholsten) [inline manifests/params.pp]: seriously, where is my head today?

File: manifests/server/config.pp [hunk 3]
```diff
@@ -117,15 +117,15 @@
   # Open read permissions to private keys to puppet group for foreman, proxy etc.
   file { "${::puppet::server::ssl_dir}/private_keys":
     ensure  => directory,
-    owner   => $::puppet::server::user,
-    group   => $::puppet::server::group,
+    owner   => $::puppet::server_private_keys_owner,
```
Comment (ekohl) [inline manifests/server/config.pp]: This (and the others in this file) should refer to ```$::puppet::server::private_keys_owner```.

File: spec/classes/puppet_server_config_spec.rb [hunk 4]
```diff
@@ -187,6 +187,18 @@
         end
       end
 
+      describe 'with server_private_keys_group' do
+        let :pre_condition do
+          'class {"puppet": server => true, server_private_keys_group => "galois"}'
+        end
+
+        it 'should set up SSL permissions' do
+          should contain_file("#{ssldir}/private_keys").with(
```
Comment (ekohl) [inline spec/classes/puppet_server_config_spec.rb]: You can also use ```.with_group('galois')``` which IMHO is slightly easier to read.

File: manifests/server/config.pp (hunk 3 above)
Comment (josephholsten) [inline manifests/server/config.pp]: changed. Though now I'm realizing I actually want to manage this directory on all nodes, not just puppetserver nodes

**Comment by mmoll** [comment]
@josephholsten You want to reuse the Puppet CA as CA for other services and it's convinient as Puppet already distributes the certs for you, that's why you want to have this on the agents, too, correct?

**Comment by josephholsten** [comment]
@mmoll yep! I've also got a related module with:

**Comment by mmoll** [comment]
@josephholsten I didn't try this, but would expect that to work (with the default of this module) if the user that needs access to the private key is in the puppet group.... at least with Puppet 3.x.

For 4.x there's no puppet group anymore so I can see your point... Did you already try out your changes in your environment?  I still think the `privatekeydir` will trigger the puppet agent on some point of the process to change the group again.

**Comment by josephholsten** [comment]
@mmoll I'm using this with 4.6.2, but there's another reason I don't want to use the puppet group: I don't want these other services to have access to everything else puppet can access. I'd be interested in middle ground, but I really only want these other services (consul, sensu &c) to have access to puppet's credentials.

**Comment by josephholsten** [comment]
really wish I'd understood what you meant by

**Comment by mmoll** [comment]
@josephholsten with the cited setting in puppet.conf, the puppet agent itself would manage the group of that dir. Maybe a similar change will be needed for the `hostprivkey` setting. But as you want to build a setup with the Puppet CA certs for other services, I guess it's easiest to test for you right now. ;)

**Comment by josephholsten** [comment]
lol. group must me `root` or `service`

**Comment by josephholsten** [comment]
I haven't got a clue where to put the `privatekeydir_group` param, so I just put it at the top. Suggestions welcome!

**Comment by josephholsten** [comment]
ok, looks like this is just not going to work. It seems that `puppet agent` runs a catalog while it's loading config. In particular, the `privatekeydir` resource sets the `group` param.

**Comment by mmoll** [comment]
@josephholsten thanks for the effort anyway! 👍

**Pull request closed without merging**